pkg/operator/encryption/state.go

```diff
@@ -122,16 +131,35 @@ func getDesiredEncryptionState(oldEncryptionConfig *apiserverconfigv1.Encryption
 	for gr, grState := range desiredEncryptionState {
 		changed := false
 		for _, expected := range expectedReadSecrets {
-			if !hasKeyInSecret(grState.readSecrets, expected) {
+			found := false
+			for _, rs := range grState.readSecrets {
+				if equalKeyInSecret(rs, expected) {
+					found = true
+					break
+				}
+			}
+			if !found {
+				// Just adding raw key without trusting any metadata on it
 				grState.readSecrets = append(grState.readSecrets, expected)
 				changed = true
 				allReadSecretsAsExpected = false
 				klog.V(4).Infof("%s missing read secret %s", gr, expected.Name)
 			}
 		}
 		if changed {
+			grState.readSecrets = sortRecentFirst(grState.readSecrets)
 			desiredEncryptionState[gr] = grState
 		}
+
+		// verify that all read secrets are backed by a real secret, and therefore
+		// in encryptionSecrets. Read secrets are sorted. So encryptionSecrets[0]
+		// is a good next write key because it is most recent.
+		for _, rs := range grState.readSecrets {
+			if !secretIsBacked(rs) {
```

Comment:
when this can happen? both `rs` and `expected` were taken from `encryptionSecrets`

Reply:
ic, https://github.com/openshift/cluster-kube-apiserver-operator/pull/603/files#diff-010b01c35a36be66869aae48f2ec96d5R91